7.18.0 blog post (#2644)
babel/website: website/blog/2022-05-18-7.18.0.md

---
layout: post
title:  "7.18.0 Released: Destructuring private elements and TypeScript 4.7"
author: Babel Team
authorURL: https://twitter.com/babeljs
date:   2022-05-18 0:00:00
categories: announcements
share_text: "Babel 7.18.0 Released"
---

We just published Babel 7.18.0!

This release includes support for the [private destructuring](https://github.com/tc39/proposal-destructuring-private) proposal and for TypeScript 4.7.

You can read the whole changelog [on GitHub](https://github.com/babel/babel/releases/tag/v7.18.0).

<!-- truncate -->

If you or your company want to support Babel and the evolution of JavaScript, but aren't sure how, you can donate to us on our [Open Collective](https://github.com/babel/babel?sponsor=1) and, better yet, work with us on the implementation of [new ECMAScript proposals](https://github.com/babel/proposals) directly! As a volunteer-driven project, we rely on the community's support to fund our efforts in supporting the wide range of JavaScript users. Reach out at [team@babeljs.io](mailto:team@babeljs.io) if you'd like to discuss more!

## Highlights

### Private destructuring ([#14304](https://github.com/babel/babel/pull/14304))

> Babel now supports transforming the [private destructuring](https://github.com/tc39/proposal-destructuring-private) proposal, which lets you use private fields or methods in destructuring patterns:

Reviewer: ```suggestion
Babel supports transforming the [private destructuring](https://github.com/tc39/proposal-destructuring-private) proposal. Now you can use private fields or methods in destructuring patterns:
```

TypeScript 4.7 supports different new syntactic features:

- instantiation expressions, which allow specifying type arguments of functions without invoking them:

  ```ts
  const identity = <T>(val: T) => val;

  const stringIdentity = identity<string>; // (val: string) => string;
  ```
- explicit variance annotations for type arguments, to guide the TypeScript type checker when computing compatibility between different types:
  ```ts
  type Provider<out T> = () => T;
  type Consumer<in T> = (x: T) => void;
  type Mapper<in T, out U> = (x: T) => U;
  type Processor<in out T> = (x: T) => T;
  ```
  In this example, `Provide<string>` is a subtype of `Provider<string | number>` while `Consumer<string | number>` is a subtype of `Consumer<string>`.
- constraints for the `infer` operator in conditional types:
  ```ts
  type GetColor<T> =
    T extends { color: infer C extends "red" | "pink" }
      ? C
      : "unknown color";
  ```

You can read the full TypeScript 4.7 release announcement [on their blog](https://devblogs.microsoft.com/typescript/announcing-typescript-4-7/).

> ### Inline `regenerator-runtime` helper support ([#14538](https://github.com/babel/babel/pull/14538))

Reviewer: ```suggestion
### Inline `regenerator-runtime` helper ([#14538](https://github.com/babel/babel/pull/14538))
```

> Starting from Babel 7.18.0, `regenerator-runtime` is automatically injected in the same way that Babel injects the other runtime helpers, without relying on an implicit `regeneratorRuntime` global. When not using [`@babel/plugin-transform-runtime`](https://babeljs.io/docs/en/babel-plugin-transform-runtime), Babel will automatically inline the `regeneratorRuntime` helper and you will not need to manually load the `regenerator-runtime` package anymore.

Reviewer: Maybe add
> you can remove "regenerator-runtime" from your package.json as Babel will automatically...

We may need to also update these docs since they are recommending direct install
- https://babeljs.io/docs/en/usage#polyfill
- https://babeljs.io/docs/en/caveats#polyfills
- https://babeljs.io/docs/en/babel-preset-env#usebuiltins-entry

Author: https://github.com/babel/website/pull/2645
